app/Foliage/PreparePackageVersion.hs

```diff
@@ -57,27 +93,67 @@ preparePackageVersion inputDir metaFile = do
   let pkgId = PackageIdentifier pkgName pkgVersion
 
   pkgSpec <-
-    readPackageVersionSpec' (inputDir </> metaFile) >>= \case
-      PackageVersionSpec {packageVersionRevisions, packageVersionTimestamp = Nothing}
-        | not (null packageVersionRevisions) -> do
-            error $
-              unlines
-                [ inputDir </> metaFile <> " has cabal file revisions but the original package has no timestamp.",
-                  "This combination doesn't make sense. Either add a timestamp on the original package or remove the revisions"
-                ]
-      PackageVersionSpec {packageVersionRevisions, packageVersionTimestamp = Just pkgTs}
-        | any ((< pkgTs) . revisionTimestamp) packageVersionRevisions -> do
-            error $
-              unlines
-                [ inputDir </> metaFile <> " has a revision with timestamp earlier than the package itself.",
-                  "Adjust the timestamps so that all revisions come after the original package"
-                ]
-      meta ->
-        return meta
+    readPackageVersionSpec' (inputDir </> metaFile) >>= \meta@PackageVersionSpec {..} -> do
+      case (NE.nonEmpty packageVersionRevisions, packageVersionTimestamp) of
+        (Just _someRevisions, Nothing) ->
+          error $
+            unlines
+              [ inputDir </> metaFile <> " has cabal file revisions but the package has no timestamp.",
+                "This combination doesn't make sense. Either add a timestamp on the original package or remove the revisions."
+              ]
+        (Just (NE.sort -> someRevisions), Just ts)
+          -- WARN: this should really be a <=
+          | revisionTimestamp (NE.head someRevisions) < ts ->
+              error $
+                unlines
+                  [ inputDir </> metaFile <> " has a revision with timestamp earlier than the package itself.",
+                    "Adjust the timestamps so that all revisions come after the package publication."
+                  ]
+          | not (null $ duplicates (revisionTimestamp <$> someRevisions)) ->
+              error $
+                unlines
+                  [ inputDir </> metaFile <> " has two revisions entries with the same timestamp.",
+                    "Adjust the timestamps so that all the revisions happen at a different time."
+                  ]
+        _otherwise -> return ()
+
+      case (NE.nonEmpty packageVersionDeprecations, packageVersionTimestamp) of
```

Comment:
lots of duplicate code??

Reply:
I tried to be smart but I failed :joy: and went back to be clear :P